content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+While we can reason about guarantees like "The system can handle 10K concurrent requests while the 99.9 percentile of requests has a latency of less than 250 ms", that's not the case for metrics like memory allocation or I/O. What does it mean if an application allocates 100 KB of RAM for a particular use case? Is it a lot? Too much? Just fine?
+* *Focused on identifying regressions:* Somewhat related to the first point, this approach of testing is focused not on specific absolute values, but rather on identifying performance regressions. It's hard to tell whether 100 KB database I/O is good or bad for a particular web request, but a change from 100 KB to 200 KB might indicate that something is wrong
+* *Focused on identifying _potential_ regressions:* A change in performance-impacting metrics does not necessarily imply an actual user-visible impacting performance regression. For instance it might be acceptable for a specific request to allocate more RAM than it did before,
+if the production system generally isn't under high load and the additional GC effort doesn't matter in practice
+* *Does not work for all performance-impacting metrics:* Some performance metrics cannot be meaningfully asserted in plain unit tests;
+e.g. degraded throughput due to lock contention can typically only be identified with a reasonable number of concurrent requests
+* *Only identifies regressions in the application itself:* A traditional integrative performance test of an enterprise application will also capture issues in related components, such as the application's database. A query run with a sub-optimal execution plan won't be noticed with this testing approach
+
+That being said, unit tests based on performance-impacting factors can be a very useful part of the performance testing strategy for an application.
+They can help to identify potential performance regressions very early in the development lifecycle,
+when they can be fixed comparatively easy and cheap.
+Of course they are no silver bullet;
+you should consider them as _complement_ for classic performance tests running on production-like hardware, not a _replacement_.
+
+The approach may feel a bit unfamiliar initially,
+and it may take some time to learn about the different metrics which can be measured with JFR and asserted via JfrUnit,
+as well as their implications on an application's performance characteristics.
+But once this hurdle is passed, continuous performance regression tests can be a valuable tool in the box of every software and performance engineer.
+
+Stay tuned for part two of this blog post, where we'll explore how to trace the SQL statements executed by an application using the JMC Agent,
+which will come in very handy for instance for identifying common performance problems like N+1 SELECT statements.
+
+_Many thanks to Hans-Peter Grahsl for his feedback while writing this blog post!_
```

Comment:
nice - but who is he? A link to him, twitter maybe?

Reply:
Yes, I'll add links to him, you, and other reviewers, unless you oppose to being mentioned :)